src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -0,0 +1,919 @@
+            |_, local| old_dataflow::DebugFormatted::new(&local),
+        );
+
+        let indirectly_mutable_locals = old_dataflow::DataflowResultsCursor::new(
+            indirectly_mutable_locals,
+            item.body,
+        );
+        let indirectly_mutable_locals = Rc::new(RefCell::new(indirectly_mutable_locals));
+
+        let needs_drop = FlowSensitiveResolver::new(
+            NeedsDrop,
+            item,
+            indirectly_mutable_locals.clone(),
+            &dead_unwinds,
+        );
+
+        let has_mut_interior = FlowSensitiveResolver::new(
+            HasMutInterior,
+            item,
+            indirectly_mutable_locals.clone(),
+            &dead_unwinds,
+        );
+
+        let qualifs = Qualifs {
+            needs_drop,
+            has_mut_interior,
+        };
+
+        Validator {
+            span: item.body.span,
+            item,
+            qualifs,
+            errors: vec![],
+            derived_from_illegal_borrow: BitSet::new_empty(item.body.local_decls.len()),
+            suppress_errors: false,
+        }
+    }
+
+    /// Resets the `QualifResolver`s used by this `Validator` and returns them so they can be
+    /// reused.
+    pub fn into_qualifs(mut self) -> Qualifs<'a, 'mir, 'tcx> {
+        self.qualifs.needs_drop.reset();
+        self.qualifs.has_mut_interior.reset();
+        self.qualifs
+    }
+
+    pub fn take_errors(&mut self) -> Vec<(Span, String)> {
+        std::mem::replace(&mut self.errors, vec![])
+    }
+
+    /// Emits an error at the given `span` if an expression cannot be evaluated in the current
+    /// context. Returns `Forbidden` if an error was emitted.
+    pub fn check_op_spanned<O>(&mut self, op: O, span: Span) -> CheckOpResult
+    where
+        O: NonConstOp + fmt::Debug
+    {
+        trace!("check_op: op={:?}", op);
+
+        if op.is_allowed_in_item(self) {
+            return CheckOpResult::Allowed;
+        }
+
+        // If an operation is supported in miri (and is not already controlled by a feature gate) it
+        // can be turned on with `-Zunleash-the-miri-inside-of-you`.
+        let is_unleashable = O::IS_SUPPORTED_IN_MIRI
+            && O::feature_gate(self.tcx).is_none();
+
+        if is_unleashable && self.tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you {
+            self.tcx.sess.span_warn(span, "skipping const checks");
+            return CheckOpResult::Unleashed;
+        }
+
+        if !self.suppress_errors {
+            op.emit_error(self, span);
+        }
+
+        self.errors.push((span, format!("{:?}", op)));
+        CheckOpResult::Forbidden
+    }
+
+    /// Emits an error if an expression cannot be evaluated in the current context.
+    pub fn check_op(&mut self, op: impl NonConstOp + fmt::Debug) -> CheckOpResult {
+        let span = self.span;
+        self.check_op_spanned(op, span)
+    }
+}
+
+impl Visitor<'tcx> for Validator<'_, 'mir, 'tcx> {
```

Comment:
This visitor is a modified version of the one in [`qualify_consts`](https://github.com/rust-lang/rust/blob/7efe1c6e678a263b9464f2e7f06f552b4d4db5a5/src/librustc_mir/transform/qualify_consts.rs#L1024). It will help to compare the two when reviewing.